### ISSUE TITLE
[kube-prometheus-stack] BadGateway in Grafana as of prometheus-operator 0.48.0

### ISSUE THREAD

**Issue by ebr** [prometheus-community/helm-charts#1013]
After upgrading `kube-prometheus-stack`, Grafana can no longer connect to Prometheus; displays `BadGateway` message. Also, I am unable to `service/kube-prometheus-kube-prome-prometheus 9090:9090` - the request times out.

Downgrading `prometheus-operator` to `v0.47.1` resolves the issue. 

```
prometheusOperator:
  image:
    repository: quay.io/prometheus-operator/prometheus-operator
    # works
    tag: v0.47.1
    # broken
    # tag: v0.48.0 
```

```
$ helm version
version.BuildInfo{Version:"v3.6.0", GitCommit:"7f2df6467771a75f5646b7f12afb408590ed1755", GitTreeState:"clean", GoVersion:"go1.16.3"}
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.0", GitCommit:"af46c47ce925f4c4ad5cc8d1fca46c7b77d13b38", GitTreeState:"clean", BuildDate:"2020-12-08T17:59:43Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19+", GitVersion:"v1.19.8-eks-96780e", GitCommit:"96780e1b30acbf0a52c38b6030d7853e575bcdf3", GitTreeState:"clean", BuildDate:"2021-03-10T21:32:29Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by LMolr** [comment]
Hi, not sure if it's the same problem as mine, I am on a fresh install from `16.1.0` helm chart.

On my setup it looks like that the `kube-prometheus-stack-prometheus` Service isn't selecting Prometheus Pods due to a Selector mismatch.

The Service has these Selectors:
- `app=prometheus`
- `prometheus=kube-prometheus-stack-prometheus`

The Prometheus Pods don't have the Label `app=prometheus`.

If I (manually) remove the `app=prometheus` Selector from the Service it starts working (checked with Grafana's 'Test Datasource').

Sadly it doesn't look like you can add Labels from the helm values (only Annotations).

I hope this helps.

**Comment by muffl0n** [comment]
Same for me. Selector in service includes `app=prometheus`
```
Name:              kube-prometheus-stack-prometheus
...
Selector:          app=prometheus,prometheus=kube-prometheus-stack-prometheus
...
```

Labels of prometheus pod on an old installation (15.2.1):
```
Name:         prometheus-kube-prometheus-stack-prometheus-0
Namespace:    monitoring
Priority:     0
Node:         foobar
Start Time:   Tue, 18 May 2021 08:54:12 +0200
Labels:       app=prometheus
              app.kubernetes.io/instance=kube-prometheus-stack-prometheus
              app.kubernetes.io/managed-by=prometheus-operator
              app.kubernetes.io/name=prometheus
              app.kubernetes.io/version=2.26.0
              controller-revision-hash=prometheus-kube-prometheus-stack-prometheus-5554bcbd5f
              operator.prometheus.io/name=kube-prometheus-stack-prometheus
              operator.prometheus.io/shard=0
              prometheus=kube-prometheus-stack-prometheus
              statefulset.kubernetes.io/pod-name=prometheus-kube-prometheus-stack-prometheus-0
...
```
And on 16.1.0:
```
Name:         prometheus-kube-prometheus-stack-prometheus-0
Namespace:    monitoring
Priority:     0
Node:         foobar
Start Time:   Thu, 27 May 2021 16:56:00 +0200
Labels:       app.kubernetes.io/instance=kube-prometheus-stack-prometheus
              app.kubernetes.io/managed-by=prometheus-operator
              app.kubernetes.io/name=prometheus
              app.kubernetes.io/version=2.27.1
              controller-revision-hash=prometheus-kube-prometheus-stack-prometheus-65ff4bf6c6
              operator.prometheus.io/name=kube-prometheus-stack-prometheus
              operator.prometheus.io/shard=0
              prometheus=kube-prometheus-stack-prometheus
              statefulset.kubernetes.io/pod-name=prometheus-kube-prometheus-stack-prometheus-0
```

**Comment by muffl0n** [comment]
Seems like this change causes the issue: https://github.com/prometheus-operator/prometheus-operator/commit/e1f4ef70818aa54641ac3a77a70a2e1469d8c927

**Comment by LMolr** [comment]
> The Prometheus Pods don't have the Label `app=prometheus`.
> 
> If I (manually) remove the `app=prometheus` Selector from the Service it starts working (checked with Grafana's 'Test Datasource').
> 
> Sadly it doesn't look like you can add Labels from the helm values (only Annotations).

After re-reading the default `values.yml` i found this (commented by default):
```yaml
    podMetadata: {}    
      # labels:
      #   app: prometheus
```

https://github.com/prometheus-community/helm-charts/blob/kube-prometheus-stack-16.1.0/charts/kube-prometheus-stack/values.yaml#L2126-L2128

It may workaround the issue for the moment..

**Comment by morremeyer** [comment]
This is a duplicate of #1009.

**Comment by MrFishFinger** [comment]
i worked around this by patching the service object:
```
kubectl --namespace=monitoring patch service kube-prometheus-stack-prometheus --type='json' -p='[{"op": "remove", "path": "/spec/selector/app"}]'
kubectl --namespace=monitoring patch service kube-prometheus-stack-prometheus -p '{"spec": {"selector": {"app.kubernetes.io/name": "prometheus"}}}'
```

**Comment by muffl0n** [comment]
This is fixed by #1010, isn't it?

**Comment by morremeyer** [comment]
@muffl0n, yes, this is fixed.
